corehq/apps/userreports/models.py

```diff
@@ -683,6 +687,48 @@ def _get_report_config(cls, static_config, domain):
         return ReportConfiguration.wrap(doc)
 
 
+class AsyncIndicator(models.Model):
+    doc_id = models.CharField(max_length=255, null=False, db_index=True)
+    doc_type = models.CharField(max_length=126, null=False)
+    domain = models.CharField(max_length=126, null=False)
+    pillow = models.CharField(max_length=126, null=False)
+    indicator_config_ids = ArrayField(
+        models.CharField(max_length=126, null=True, blank=True),
+        null=False
+    )
+    date_created = models.DateTimeField(auto_now_add=True, db_index=True)
+
+    class Meta(object):
+        unique_together = ('doc_id', 'pillow',)
+        ordering = ["date_created"]
+
+    @classmethod
+    def update_indicators(cls, change, pillow, table_ids):
+        doc_id = change.id
+        pillow_id = pillow.pillow_id
+        with CriticalSection([get_async_indicator_modify_lock_key(doc_id, pillow_id)]):
+            try:
+                indicator = cls.objects.get(doc_id=doc_id, pillow=pillow_id)
+            except cls.DoesNotExist:
+                doc_type = change.document['doc_type']
+                domain = change.document['domain']
+                indicator = AsyncIndicator(
+                    doc_id=doc_id,
+                    doc_type=doc_type,
+                    domain=domain,
+                    pillow=pillow_id,
+                    indicator_config_ids=table_ids
+                )
+            else:
+                current_table_ids = indicator.indicator_config_ids
+                new_table_ids = list(set(current_table_ids).union(set(table_ids)))
+                indicator.indicator_config_ids = new_table_ids
```

Comment:
If the indicator is already queued up to be put into celery, then this will add any new UCR data sources to that indicator